### PR TITLE
Closes #59 - Extending configuration to enable an annotation based target selection

### LINKED ISSUE #59: Extending configuration to enable an annotation based target selection
Should support:

* interface
* super-class
* class-annotation
* method-annotation
* method parameters (by default all methods with given name)
* modifiers (only public by default? at least private should be not instrumented by default)

Please separate the class name, interface and super-class, we should allow configs that define all 3 and not like now only one FQN.

Nice to have (can be separated into further tickets and moved to another milestone):
* wild-card (this is great, but we need to be careful here for what we want a wild-card to be possible)
  * don't allow `*` only
* when the instrumenting by interface, instrument only methods available in that interface
* when the instrumenting by super-class, instrument only methods available in that super-class

### PR DESCRIPTION
**Pull request by mariusoe**

### CODE REVIEW

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/config/InstrumentationScopeResolver.java [hunk 1]
```diff
@@ -94,33 +95,31 @@ private InstrumentationScope resolveScope(InstrumentationScopeSettings scopeSett
...
-     * Creating a matcher for the given {@link NameMatcherSettings} which represents concrete classes.
+     * Creating and adding a matcher for the given {@link NameMatcherSettings} which represents concrete classes.
      */
-    private void processType(MatcherChainBuilder<TypeDescription> builder, NameMatcherSettings nameSettings) {
-        ElementMatcher.Junction<TypeDescription> nameMatcher = nameIs(nameSettings);
-        if (nameMatcher != null) {
-            builder.and(nameMatcher);
-        }
+    private void processType(MatcherChainBuilder<TypeDescription> builder, DescriptionMatcherSettings descriptionSettings) {
+        ElementMatcher.Junction<TypeDescription> nameMatcher = describedBy(descriptionSettings);
```
Comment (JonasKunz) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/config/InstrumentationScopeResolver.java]: Maybe also rename the variable "nameMatcher" here, as it does not only match by name anymore

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/DispatchHookAdvice.java [hunk 2]
```diff
@@ -14,8 +14,8 @@
      * The content of this method is added before the target method! See {@link Advice} for existing parameter annotations.
      */
     @Advice.OnMethodEnter
-    public static void onEnter(@Advice.Origin Object origin) {
-        System.out.println("\tEntering: " + origin);
+    public static void onEnter(@Advice.Origin Object origin, @Advice.This Object thiz) {
```
Comment (JonasKunz) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/DispatchHookAdvice.java]: this still also works on static methods, right?

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettingsTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        @Test
+        public void allAnnotationMatcher() {
+            DescriptionMatcherSettings settings = new DescriptionMatcherSettings();
+            settings.setMatcherMode(StringMatcher.Mode.STARTS_WITH);
+            NameMatcherSettings matcher = new NameMatcherSettings();
+            matcher.setMatcherMode(StringMatcher.Mode.STARTS_WITH);
+            settings.setAnnotations(Collections.singletonList(matcher));
+
+            boolean result = settings.isAnyMatcher();
```
Comment (JonasKunz) [inline inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettingsTest.java]: why is this an any-matcher? It only matches types which have at least one annotation, doesn't it?

File: .gitignore [hunk 4]
```diff
@@ -1,6 +1,7 @@
 .gradle
 *build/
 !gradle/wrapper/gradle-wrapper.jar
+*.log
```
[the remark this answers is not in the document]
Comment (mariusoe) [reply .gitignore]: Yep :)

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettings.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.List;
+
+/**
+ * Extending the {@link NameMatcherSettings} by adding the ability to specify a list of {@link NameMatcherSettings} which
+ * represents annotations.
+ */
+@Data
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
+public class DescriptionMatcherSettings extends NameMatcherSettings {
```
[the remark this answers is not in the document]
Comment (mariusoe) [reply inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettings.java]: That was the best name I had :D but I'll go for the `ElementDescriptionMatcherSettings` your proposed

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/config/InstrumentationScopeResolver.java [hunk 6]
```diff
@@ -94,33 +95,31 @@ private InstrumentationScope resolveScope(InstrumentationScopeSettings scopeSett
...
     /**
-     * Creating a matcher for the given {@link NameMatcherSettings} which represents a superclass.
+     * Creating and adding a matcher for the given {@link NameMatcherSettings} which represents a superclass.
      */
-    private void processSuperclass(MatcherChainBuilder<TypeDescription> builder, NameMatcherSettings nameSettings) {
-        ElementMatcher.Junction<NamedElement> nameMatcher = nameIs(nameSettings);
-        if (nameMatcher != null) {
-            builder.and(hasSuperType(not(isInterface()).and(nameMatcher)));
+    private void processSuperclass(MatcherChainBuilder<TypeDescription> builder, DescriptionMatcherSettings descriptionSettings) {
+        ElementMatcher.Junction<TypeDescription> matcher = describedBy(descriptionSettings);
```
[the remark this answers is not in the document]
Comment (mariusoe) [reply inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/config/InstrumentationScopeResolver.java]: Using `ElementDescriptionMatcherSettings`

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/DispatchHookAdvice.java (hunk 2 above)
Comment (mariusoe) [reply inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/DispatchHookAdvice.java]: No it doesn't :) this was a test and shouldn't be checked in

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettingsTest.java (hunk 3 above)
Comment (mariusoe) [reply inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/config/model/instrumentation/scope/DescriptionMatcherSettingsTest.java]: You're right. This case shouldn't be a problem - changed it.

**Pull request merged**